### PR TITLE
[exporter/kafka] Add factory options for custom metric and log marshalers

### PR DESCRIPTION
**Pull request by bgranetzke**
Description:
Adding ExportFactory options to add custom log/metric marshalers (similar to existing: WithTracesMarshalers)

Link to tracking Issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14514

Testing:
Tests were added for the new options as well as adapting the previous test for WithTracesMarshalers to the new method.

### CODE REVIEW

**Comment by bgranetzke** [comment]
@MovieStoreGuy Sorry for creating a new PR.  This should have all your recommendations incorporated.  It also leverages your latest factory_test approach.

File: exporter/kafkaexporter/factory_test.go [hunk 1]
```diff
@@ -155,10 +217,14 @@ func TestCreateLogExporter(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			t.Parallel()
 
-			f := NewFactory()
+			core, observed := observer.New(zapcore.DebugLevel)
+			exporterSettings := componenttest.NewNopExporterCreateSettings()
+			exporterSettings.Logger = zap.New(core)
```
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/factory_test.go:222]: I am still not a fan of using log statements to validate a setting, what I suggest you do is use the mock marshaler and call the exporter.

File: exporter/kafkaexporter/go.mod [hunk 2]
```diff
@@ -22,6 +22,7 @@ require (
 
 require (
 	github.com/apache/thrift v0.17.0 // indirect
+	github.com/benbjohnson/clock v1.3.0 // indirect
```
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/go.mod:25]: Did this come up in a `go mod tidy` ?

File: exporter/kafkaexporter/factory_test.go (hunk 1 above)
Comment (bgranetzke) [reply exporter/kafkaexporter/factory_test.go:222]: Then the choices are to either add a config setting that tells the producer to noop or abstract the producer such that we can inject a noop impl for the test ([my original PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13433)).  Do you have a preference?

File: exporter/kafkaexporter/go.mod (hunk 2 above)
Comment (bgranetzke) [reply exporter/kafkaexporter/go.mod:25]: I swear `make gotidy` added it, but now it doesn't.  I'll make sure it's pulled in the next iteration.

Comment (MovieStoreGuy) [reply exporter/kafkaexporter/go.mod:25]: Not to worry, `go mod tidy` does so magic at times.

File: exporter/kafkaexporter/factory_test.go (hunk 1 above)
Comment (MovieStoreGuy) [reply exporter/kafkaexporter/factory_test.go:222]: Right, so it took me a bit to figure out why I wasn't certain on first PR. 

I would rather a safe default, instead nil implying default.

So a Noop type following what I had done here https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/awskinesisexporter/exporter.go#LL56 would be perfect if you have the time.

Comment (MovieStoreGuy) [reply exporter/kafkaexporter/factory_test.go:222]: It looks like my PR got merged in thinking yours got stale.

Comment (bgranetzke) [reply exporter/kafkaexporter/factory_test.go]: Removed log-based validation of marshaler

**Comment by bgranetzke** [comment]
@MovieStoreGuy It was a bit more involved than initially thought.  I've removed the log-based assertion of the expected marshaler being selected and replaced with an approach that validates the marshaler by assessing the exporter output. 
 I'd love to get your feedback.  Callouts:
1. re-introduced the ProducerFactoryFunc from the original PR.  I believe the existing sarama.SyncProducer interface (along with the equivalent sarama/mocks.SyncProducer) is sufficient for testing without introducing a new abstraction.
2. addressed your original defaulting concern (vs. defaulting to nil)
3. Adapted your test cases to include producerFunc in the test scenario structs

**Comment by MovieStoreGuy** [comment]
Let me have a further look to see how i can help :)

**Comment by MovieStoreGuy** [comment]
Once you get a chance @bgranetzke , please rebase :)

**Comment by bgranetzke** [comment]
@MovieStoreGuy Thank you for the changes.  Correct me if I've missed something, but there is no longer a reason for this PR, correct?  Just checking before I close it.  Thanks again!

**Comment by MovieStoreGuy** [comment]
I don't believe so @bgranetzke , you should be safe to close this out.

**Pull request closed without merging**